### PR TITLE
[SPARK-31337][SQL]Support MS SQL Kerberos login in JDBC connector

### PR DESCRIPTION
**Pull request by gaborgsomogyi**
### What changes were proposed in this pull request?
When loading DataFrames from JDBC datasource with Kerberos authentication, remote executors (yarn-client/cluster etc. modes) fail to establish a connection due to lack of Kerberos ticket or ability to generate it.

This is a real issue when trying to ingest data from kerberized data sources (SQL Server, Oracle) in enterprise environment where exposing simple authentication access is not an option due to IT policy issues.

In this PR I've added MS SQL support.

What this PR contains:
* Added `MSSQLConnectionProvider`
* Added `MSSQLConnectionProviderSuite`
* Changed MS SQL JDBC driver to use the latest (test scope only)
* Changed `MsSqlServerIntegrationSuite` docker image to use the latest
* Added a version comment to `MariaDBConnectionProvider` to increase trackability

### Why are the changes needed?
Missing JDBC kerberos support.

### Does this PR introduce _any_ user-facing change?
Yes, now user is able to connect to MS SQL using kerberos.

### How was this patch tested?
* Additional + existing unit tests
* Existing integration tests
* Test on cluster manually

### CODE REVIEW

**Comment by gaborgsomogyi** [comment]
Apart from manual testing I've tried to add docker integration test (failed) and tried out the following:
* Set up an active directory docker => not yet supported, please see [this](https://github.com/microsoft/mssql-docker/issues/165) link for details.
* Use Hadoop `MiniKdc` => simply never authenticated, furthermore MS never claimed it is working

If there will be a working docker image with an Active Directory instance we can try to it again. In the meantime if somebody has an idea how to overcome this feel free to add.

File: external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala [hunk 1]
```diff
@@ -27,7 +27,7 @@ import org.apache.spark.tags.DockerTest
 @DockerTest
 class MsSqlServerIntegrationSuite extends DockerJDBCIntegrationSuite {
   override val db = new DatabaseOnDocker {
-    override val imageName = "mcr.microsoft.com/mssql/server:2017-GA-ubuntu"
+    override val imageName = "mcr.microsoft.com/mssql/server:2019-GA-ubuntu-16.04"
```
Comment (gaborgsomogyi) [inline external/docker-integration-tests/src/test/scala/org/apache/spark/sql/jdbc/MsSqlServerIntegrationSuite.scala:30]: This is not absolutely necessary, if we think we can extract it into a new PR. Thought it would be overkill.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+private[sql] class MSSQLConnectionProvider(
+    driver: Driver,
+    options: JDBCOptions,
+    parserMethod: String = "parseAndMergeProperties"
+  ) extends SecureConnectionProvider(driver, options) {
+  override val appEntry: String = {
+    val configName = "jaasConfigurationName"
+    val appEntryDefault = "SQLJDBCDriver"
+
+    val parseURL = try {
```
Comment (gaborgsomogyi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala:37]: There are basically 2 approaches to parse the URL to get `jaasConfigurationName`:
* Try to call private [parseAndMergeProperties](https://github.com/microsoft/mssql-jdbc/blob/0d4e97f401dc0e55779460d9709dd7ee399246be/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDriver.java#L831-L854) => tried it first
* Parse the URL manually => used as fallback

Both way has been tested in `MSSQLConnectionProviderSuite`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+import java.security.PrivilegedExceptionAction
+import java.sql.{Connection, Driver}
+import java.util.Properties
+
+import org.apache.hadoop.security.UserGroupInformation
+
+import org.apache.spark.sql.execution.datasources.jdbc.JDBCOptions
+
+private[sql] class MSSQLConnectionProvider(
```
Comment (gaborgsomogyi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala:28]: The implementation is based on [this](https://docs.microsoft.com/en-us/sql/connect/jdbc/using-kerberos-integrated-authentication-to-connect-to-sql-server?view=sql-server-ver15).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+          MSSQLConnectionProvider.super.getConnection()
+        }
+      }
+    )
+  }
+
+  override def getAdditionalProperties(): Properties = {
+    val result = new Properties()
+    result.put("integratedSecurity", "true")
+    result.put("authenticationScheme", "JavaKerberos")
```
Comment (gaborgsomogyi) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala:73]: These props needed to reach kerberos authentication case:
https://github.com/microsoft/mssql-jdbc/blob/0d4e97f401dc0e55779460d9709dd7ee399246be/src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java#L3771-L3772

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProviderSuite.scala [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    val defaultProvider = new MSSQLConnectionProvider(
+      driver, options("jdbc:sqlserver://localhost/mssql"))
+    val customProvider = new MSSQLConnectionProvider(
+      driver, options(s"jdbc:sqlserver://localhost/mssql;jaasConfigurationName=custommssql"))
+
+    testProviders(defaultProvider, customProvider)
+  }
+
+  test("setAuthenticationConfigIfNeeded custom parser must set authentication if not set") {
+    val parserMethod = "IntentionallyNotExistingMethod"
```
Comment (gaborgsomogyi) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProviderSuite.scala:32]: This simulates a driver where the private parser method doesn't exist. Such case manual parsing takes place and the code goes forward without issue.

**Review by gaborgsomogyi — COMMENTED** [review]
Couple of pointers to help reviewers.

There is a driver wrapper [here](https://github.com/nabacg/krb5sqljdb/blob/master/src/main/scala/net/cabworks/jdbc/Krb5SqlServer.scala) which achieves more or less the same behaviour but with an external jar. The blogpost can be found [here](https://datamountaineer.com/2016/01/15/spark-jdbc-sql-server-kerberos/).

**Comment by gaborgsomogyi** [comment]
cc @HeartSaVioR

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProviderSuite.scala [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    val driver = registerDriver(MSSQLConnectionProvider.driverClass)
+    val defaultProvider = new MSSQLConnectionProvider(
+      driver, options("jdbc:sqlserver://localhost/mssql"), parserMethod)
+    val customProvider = new MSSQLConnectionProvider(
+      driver,
+      options(s"jdbc:sqlserver://localhost/mssql;jaasConfigurationName=custommssql"),
```
Comment (seanli-rallyhealth) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProviderSuite.scala:38]: no need of s String unless there is embedded variable

Comment (gaborgsomogyi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProviderSuite.scala:38]: Fixed.

File: dev/deps/spark-deps-hadoop-3.2-hive-2.3 [hunk 7]
```diff
@@ -183,7 +183,6 @@ metrics-jmx/4.1.1//metrics-jmx-4.1.1.jar
 metrics-json/4.1.1//metrics-json-4.1.1.jar
 metrics-jvm/4.1.1//metrics-jvm-4.1.1.jar
 minlog/1.3.0//minlog-1.3.0.jar
-mssql-jdbc/6.2.1.jre7//mssql-jdbc-6.2.1.jre7.jar
```
Comment (dongjoon-hyun) [inline dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: Oh, interesting. I didn't notice this.
@gaborgsomogyi . This is worth for minimizing Apache Spark 3.0.0 distribution.
Since this PR cannot land on `branch-3.0`, could you make another PR for this dependency clean-up?

Comment (dongjoon-hyun) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: Do you know what was the root cause of this dependency?

Comment (gaborgsomogyi) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: I've just double checked and it has been added by a guy at 24 Nov 2019 [here](https://github.com/apache/spark/blob/a1706e2fa745a8d4aebeebd1663366e13ba80c8c/dev/deps/spark-deps-hadoop-3.2-hive-2.3) :)
You can ask him what was the reason :D

Comment (gaborgsomogyi) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: > could you make another PR for this dependency clean-up?

Sure. Creating a jira + PR...

Comment (gaborgsomogyi) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: Filed https://issues.apache.org/jira/browse/SPARK-31821

Comment (gaborgsomogyi) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: Thanks @dongjoon-hyun Going to roll back the dep remove here as soon as SPARK-31821 is merged.

Comment (dongjoon-hyun) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: I know you are joking. ;) But, that is an auto-generated file.
> I've just double checked and it has been added by a guy at 24 Nov 2019

Comment (dongjoon-hyun) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: The PR failed at https://github.com/apache/spark/pull/28640#issuecomment-633919385 .

Comment (gaborgsomogyi) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3:186]: > But, that is an auto-generated file.

Now I see :) Just fixed the other PR.

Comment (gaborgsomogyi) [reply dev/deps/spark-deps-hadoop-3.2-hive-2.3]: Merged master into this branch to eliminate the deps change. Thanks @dongjoon-hyun

**Comment by gaborgsomogyi** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+      m.setAccessible(true)
+      Some(m)
+    } catch {
+      case _: NoSuchMethodException => None
+    }
+
+    parseURL match {
+      case Some(m) =>
+        logDebug("Property parser method found, using it")
+        m.invoke(driver, options.url, null).asInstanceOf[Properties]
```
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala:48]: Let's leave method signature as comment - it makes easier to verify when the target method to call should be changed.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala (hunk 4 above)
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala:73]: It'd be nice to add the background information on comment, either the code side or the class doc.

**Review by HeartSaVioR — COMMENTED** [review]
I've gone through the general review but I'm afraid I'm not qualified to review in details.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala (hunk 8 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala:50]: Added.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala (hunk 4 above)
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/connection/MSSQLConnectionProvider.scala:76]: Added an inline comment.

**Comment by gaborgsomogyi** [comment]
cc @vanzin

**Comment by vanzin** [comment]
retest this please

**Comment by vanzin** [comment]
Merging to master.

**Comment by thereverand** [comment]
I could be wrong, but it doesn't look like this fixes this issue completely.
It doesn't support username/password and Kerberos, which the driver supports, and if the proper krb5.conf and JAAS config are in place does work on the driver node of a Databricks cluster, but then fails on the remote executors in the cluster with the same error being described here. So it looks to me to be the same issue.

**Comment by gaborgsomogyi** [comment]
@thereverand just seen your comment. It has never stated that username/password is working.
Propagating u/p through the whole cluster will enlarge the attack surface.
This supports keytab which can be better protected.

As a resolution we're planning to add JDBC connection provider API in SPARK-32001 where any custom provider can be added.
This allows each provider to add custom authentication possibilities.
